DMD.C

```diff
@@ -0,0 +1,489 @@
+    delete d_phi_imaginary_squared_2;
+    delete  d_projected_init_imaginary_1;
+    delete  d_projected_init_imaginary_2;
+    delete rhs_real;
+    delete rhs_imaginary;
+
+    delete [] inverse_input;
+    delete [] ipiv;
+    delete [] work;
+}
+
+Vector*
+DMD::predict(double t,
+             double dt)
+{
+    const std::pair<Vector*, Vector*> d_projected_init_pair(d_projected_init_real, d_projected_init_imaginary);
+    return predict(d_projected_init_pair, t, dt);
+}
+
+Vector*
+DMD::predict(const std::pair<Vector*, Vector*> init,
+             double t,
+             double dt)
+{
+    std::pair<Matrix*, Matrix*> d_phi_pair = phiMultEigs(t, dt);
+    Matrix* d_phi_mult_eigs_real = d_phi_pair.first;
+    Matrix* d_phi_mult_eigs_imaginary = d_phi_pair.second;
+
+    Vector* d_predicted_state_real_1 = d_phi_mult_eigs_real->mult(init.first);
+    Vector* d_predicted_state_real_2 = d_phi_mult_eigs_imaginary->mult(init.second);
+    Vector* d_predicted_state_real = d_predicted_state_real_1->minus(d_predicted_state_real_2);
+
+    delete d_phi_mult_eigs_real;
+    delete d_phi_mult_eigs_imaginary;
+    delete d_predicted_state_real_1;
+    delete d_predicted_state_real_2;
+
+    return d_predicted_state_real;
+}
+
+std::pair<Matrix*, Matrix*>
+DMD::phiMultEigs(double t,
+                 double dt)
+{
+    Matrix* d_eigs_exp_real = new Matrix(d_k, d_k, false);
+    Matrix* d_eigs_exp_imaginary = new Matrix(d_k, d_k, false);
+
+    double exp_power = t / dt;
+
+    for (int i = 0; i < d_k; i++)
+    {
+        std::complex<double> eig_exp = std::pow(d_eigs[i], exp_power);
+        d_eigs_exp_real->item(i, i) = std::real(eig_exp);
+        d_eigs_exp_imaginary->item(i, i) = std::imag(eig_exp);
+    }
+
+    Matrix* d_phi_mult_eigs_real = d_phi_real->mult(d_eigs_exp_real);
+    Matrix* d_phi_mult_eigs_real_2 = d_phi_imaginary->mult(d_eigs_exp_imaginary);
+    *d_phi_mult_eigs_real -= *d_phi_mult_eigs_real_2;
+    Matrix* d_phi_mult_eigs_imaginary = d_phi_real->mult(d_eigs_exp_imaginary);
+    Matrix* d_phi_mult_eigs_imaginary_2 = d_phi_imaginary->mult(d_eigs_exp_real);
+    *d_phi_mult_eigs_imaginary += *d_phi_mult_eigs_imaginary_2;
+
+    delete d_eigs_exp_real;
+    delete d_eigs_exp_imaginary;
+    delete d_phi_mult_eigs_real_2;
+    delete d_phi_mult_eigs_imaginary_2;
+
+    return std::pair<Matrix*,Matrix*>(d_phi_mult_eigs_real, d_phi_mult_eigs_imaginary);
+}
+
+const Matrix*
+DMD::getSnapshotMatrix()
+{
+    CAROM_VERIFY(d_snapshots.size() > 0);
+    CAROM_VERIFY(d_snapshots[0].dim() > 0);
+    for (int i = 0 ; i < d_snapshots.size() - 1; i++)
+    {
+        CAROM_VERIFY(d_snapshots[i].dim() == d_snapshots[i + 1].dim());
+        CAROM_VERIFY(d_snapshots[i].distributed() == d_snapshots[i + 1].distributed());
+    }
+
+    Matrix* snapshot_mat = new Matrix(d_snapshots[0].dim(), d_snapshots.size(), d_snapshots[0].distributed());
+
+    // This might be slow since we're accessing a different memory address each time.
```

Comment:
Is it clear from testing whether this is a bottleneck, or would timing need to be done for some large problems?

Reply:
Good timing. I just ran into an issue with Youngsoo with slow DMD. This is not a bottleneck. It is quite fast. The bottleneck is the scalapack SVD call.